lib/ddtrace/sampling/rule_sampler.rb

```diff
@@ -0,0 +1,56 @@
+require 'ddtrace/ext/priority'
+
+require 'ddtrace/ext/sampling'
+require 'ddtrace/sampler'
+
+module Datadog
+  module Sampling
+    # TODO:
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :priority_sampler
+
+      def initialize(rules, rate_limiter, priority_sampler = Datadog::RateByServiceSampler.new)
+        @rules = rules
+        @rate_limiter = rate_limiter
+        @priority_sampler = priority_sampler
+      end
+
+      def sample?(span)
+        sampled, _ = sample_span(span) { |s| return @priority_sampler.sample?(s) }
+        sampled
+      end
+
+      def sample!(span)
+        sampled, sample_rate = sample_span(span) { |s| return @priority_sampler.sample!(s) }
+
+        # Set metrics regardless of sampling outcome
+        set_metrics(span, sample_rate, rate_limiter.effective_rate)
+
+        span.sampled = sampled
+        sampled
+      end
+
+      def_delegators :@priority_sampler, :update
+
+      private
+
+      def sample_span(span)
+        sampled, sample_rate = @rules.find do |rule|
```

Comment:
This loop is a little awkward because you have to find and retrieve two different values from the rule. It may be possible to eliminate `Rule#sample` entirely, and simplify things. If we were to change `Rule` to delegate to `Matcher#match?`, `Sampler#sample?`, `Sampler#sample_rate`, then you could rewrite this function as:

```ruby
def sample_span(span)
  rule = @rules.find do |rule|
    rule.matches?(span)
  end

  yield(span) unless rule

  sampled = rule.sample?(span)
  sample_rate = rule.sample_rate(span)
  [sampled && rate_limiter.allow?(1), sample_rate]
end
```

As a general rule of thumb, it might be better for components like `Rule` to have very simple methods that do as little as possible and have more complex, compositional objects such as this `RuleSampler` orchestrate and drive those smaller components.

There might be even better ways of doing this than what I've suggested, but just some food for thought.

Reply:
There is one trade-off with this approach which is that custom rules might want to provide a coupled result between the sampling decision (`sample?`) and sample_rate. In this case we are making it harder to implement such rules:
```ruby
class CustomRule < Rule
  def sample?(span)
    case span.service
    when 'service-1'
      span.name != 'ignore.op'
    when 'service-2'
      span.start_time.hour > 7
    end
  end

  def sample_rate(span)
    case span.service
    when 'service-1'
      0.8
    when 'service-2'
      0.9
    end
  end
end
```

Instead of returning a "response" payload with both together:
```ruby
Rule.new do |span|
  case span.service
  when 'service-1'
    [span.name != 'ignore.op', 0.8]
  when 'service-2'
    [span.start_time.hour > 7, 0.9]
  end
end
```
After reviewing this, it seems like an ergonomic decision that we wouldn't prioritize in detriment of the maintainability of `RuleSampler`. Thus I made changes that mostly follow your suggestion.